lib/ASTGen/Sources/ASTGen/Generics.swift

```diff
@@ -87,4 +89,117 @@ extension ASTGenVisitor {
+
+      case .prefixOperatorExpr(let node):
+        let op = node.operator
+
+        guard op.text == "-" else {
+          fatalError("Unknown prefix operator for same type left type")
+        }
+
+        guard case .integerLiteralExpr(let node) = node.expression.as(ExprSyntaxEnum.self) else {
+          fatalError("Unknown expression kind for same type left type")
+        }
+
+        return self.generate(integerType: node, minusLoc: self.generateSourceLoc(op))
+
+      default:
+        fatalError("Unknown expression kind for same type left type")
+      }
+    }
+  }
+
+  func generate(sameTypeRightType node: SameTypeRequirementSyntax.RightType) -> BridgedTypeRepr {
+    switch node {
+    case .type(let type):
+      return self.generate(type: type)
+
+    // The only expressions same type right types support right now are
+    // integer literals, '123', and prefix operators for negative integer
+    // literals, '-123'.
+    case .expr(let expr):
+      switch expr.as(ExprSyntaxEnum.self) {
+      case .integerLiteralExpr(let node):
+        return self.generate(integerType: node)
+
+      case .prefixOperatorExpr(let node):
+        let op = node.operator
+
+        guard op.text == "-" else {
+          fatalError("Unknown prefix operator for same type right type")
+        }
+
+        guard case .integerLiteralExpr(let node) = node.expression.as(ExprSyntaxEnum.self) else {
+          fatalError("Unknown expression kind for same type right type")
+        }
+
+        return self.generate(integerType: node, minusLoc: self.generateSourceLoc(op))
+
+      default:
+        fatalError("Unknown expression kind for same type right type")
+      }
+    }
+  }
+
+  func generate(genericArgument node: GenericArgumentSyntax.Argument) -> BridgedTypeRepr {
+    switch node {
+    case .type(let type):
+      return self.generate(type: type)
+
+    // The only expressions generic argument types support right now are
+    // integer literals, '123', and prefix operators for negative integer
+    // literals, '-123'.
+    case .expr(let expr):
+      switch expr.as(ExprSyntaxEnum.self) {
+      case .integerLiteralExpr(let node):
+        return self.generate(integerType: node)
+
+      case .prefixOperatorExpr(let node):
+        let op = node.operator
+
+        guard op.text == "-" else {
+          fatalError("Unknown prefix operator for generic argument type")
+        }
+
+        guard case .integerLiteralExpr(let node) = node.expression.as(ExprSyntaxEnum.self) else {
+          fatalError("Unknown expression kind for generic argument type")
+        }
+
+        return self.generate(integerType: node, minusLoc: self.generateSourceLoc(op))
+
+      default:
+        fatalError("Unknown expression kind for generic argument type")
+      }
+    }
+  }
+
+  func generate(
+    integerType node: IntegerLiteralExprSyntax,
+    minusLoc: BridgedSourceLoc = BridgedSourceLoc()
+  ) -> BridgedTypeRepr {
```

Comment:
```suggestion
  func generateIntegerType(
    integerLiteralExpr node: IntegerLiteralExprSyntax,
    minusLoc: BridgedSourceLoc = BridgedSourceLoc()
  ) -> IntegerLiteralTypeRepr {
```

ASTGen has a loose `generate` function naming convention `generate{Result}({source}:)` where `{Result}` is what to generate, and `{source}` is the input syntax type. Also we'd like to delay upcasting (`.asTypeRepr`) as much as possible.

Also could this function include the common logic between `case .expr(let expr):` bodies of the two functions above? i.e. 
```swift
  func generateIntegerType(expr node: Expr) -> BridgedIntegerTypeRepr {
    let minusLoc: SourceLoc
    let literalExpr: integerLiteralExprSyntax

    switch expr.as(ExprSyntaxEnum.self) {
    case .integerLiteralExpr(let node):
      minusLoc = nil
      literalExpr = node
    case .prefixOperatorExpr(let node):
      // .. check `-{IntergerLiteralSyntax}` ...
      minusLoc = self.generateSourceLoc(op)
      literalExpr = ...
    default:
        fatalError(...)
    }

    return .createParsed(
      self.ctx,
      string: self.copyAndStripUnderscores(text: literalExpr.rawText),
      loc: self.generateSourceLoc(literalExpr),
      minusLoc: minusLoc
    )
  }
```